### DISCUSSION TITLE
Links and self references

### DISCUSSION THREAD

**Discussion by h34lthbybr33** [avitorio/outstatic#256] [Ideas]
Searched a bit and didn't find it, but is there a way link refinement can be put on the roadmap? To explain what I mean:

This content is usually driving another facet of the site (either a static site builder, website, etc.) Each of those should have a "home" on the finished site. It would be nice if, through a metadata component, you could assign how each collection manifests/is published, then back-reference that when creating links in documents.

For example, if you have a "Posts" collection, and a document with "hello-world" slug, it would be nice (if at the collection level) you could define the final URL as `/blog/${slug}`, then within a document when you're inserting a link, you can either provide an absolute link, or point to an item from a collection (then it builds it replacing necessary tokens).

This doesn't have to get so intricate where you maintain a links database and update every document using that reference should it change (though that'd be nice as well), but would make it easy to reference pre-existing content. (Side note, that same links database could also be expanded/exposed for "related posts, in this example).

Hope that makes sense, but LMK if I need to give a better explain/more concise definition.

**Comment by avitorio** [comment]
Hey @h34lthbybr33 ,

Thanks for the suggestion. This would be a cool feature, but I think somewhat complex to achieve.

I like the idea of suggesting documents when creating links. I'm just wondering where we would pull that info from. 
Would we add that info as metadata to each document or on a separate json file? Should we create a collection wide base url?
How do we fetch that info?  

I think we will add this feature at some point and I'm happy to add it to the roadmap/backlog, but this one would require some investigation.

It would be great if you could think about how we can implement this and what the UX would look like. I'm open to suggestions!

**Reply by h34lthbybr33** [reply]
In my eyes, collection level config decides how that document is exposed (or if it's exposed, as these could just drive components, too).
The `schema.json` contains the URL "template" (where you can insert field values as tokens):

```
{
  "title": "Posts",
  "type": "object",
  "properties": { ... },
  "exposedUrl": "/blog/%slug%"
}
```

Then at document save you build the URL into the final `metadata.json`:

```
{
  "commit": "...",
  "generated": "2024-10-23T12:41:33.680Z",
  "metadata": [
    {
      "__outstatic": {
        "commit": "...",
        "hash": "...",
        "path": ".../posts/hello-world.md"
        "exposedUrl": "/blog/hello-world"
      }
    }
  ]
}
```

I'll also call out that if/when that document 'exposedUrl' is updated/changed, the existing files get re-iterated, and the changes manifest back up to metadata (like a chore of sorts).

Then the link popup, in my eyes, stays consolidated but with the ability to select a document, too. I could see it at least a couple of different ways. One option, a CTA within the existing link component. Maybe something like:
![image](https://github.com/user-attachments/assets/4037367f-1a69-4eb1-86be-437e4ce0e772)
Another option is add a new button to format menu, one that makes it more concise it's a document reference and not just broadly a link:

![image](https://github.com/user-attachments/assets/fbad2815-0788-4f51-82e2-9fa7349008d5)

As for selection, just expose what is "link-able" (based on that "exposedUrl") and allow selection:

![image](https://github.com/user-attachments/assets/07837e5a-4bad-4260-bf0f-e8e4a7020c0c)

The search could be in another iteration as well. I think MVP is just the metadata/schema and a way to expose those documents that support it.

To that point, this addition also allows you to entertain things like "view draft" on the document edit page. Somewhere in this space, maybe another icon beside trash?

![image](https://github.com/user-attachments/assets/164efc83-50e1-47ba-af01-b29fbbeaf2b8)

BTW, excuse my "wireframe", lol. But it gets the concept across (I hope).

**Reply by avitorio** [reply]
Thank you for the detailed response. This is an excellent solution that I've added to our backlog with a reference to this discussion. While there are higher-priority features ahead of it, we'll definitely implement this enhancement in the future. I'll keep you updated on our progress.
